### PR TITLE
Todo List with Redux

### PR DESCRIPTION
**Pull request by mtalha-mergestack**
features:
- TypeScript Implemented
- Redux Implemented
- Containerized the app

### CODE REVIEW

File: src/components/TodoItem/TodoItem.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  const editTask = (key: string, index: number, value: string): void => {
+    if (key == "Enter" && isCompleted != true) {
+      editListHandler(index, value);
+      setIsEditing(false);
+    }
+  };
+
+  const completed = (isChecked: boolean): void => {
+    isChecked ? setIsCompleted(true) : setIsCompleted(false);
+  };
```
Comment (danialsyed6) [inline src/components/TodoItem/TodoItem.tsx]: This is just extra code. You could just do `setIsCompleted(event.target.checked)` and it should work as checked is also a boolean.

File: src/components/TodoItem/TodoItem.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          />
+        ) : (
+          value
+        )}
+      </span>
+      {isCompleted || (
+        <i
+          onClick={() => {
+            setIsEditing(!isEditing);
+          }}
```
Comment (danialsyed6) [inline src/components/TodoItem/TodoItem.tsx]: Use single line arrow function

File: src/store/reducers.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+interface TodoState {
+  todos: string[];
+}
```
Comment (danialsyed6) [inline src/store/reducers.ts]: How is todo an array of strings? Doesn’t it have isCompleted state in it? It should be an object with an id.

File: src/store/reducers.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+const initialState: TodoState = {
+  todos: [],
+};
+interface RootState {
+  todo: TodoState;
+}
```
Comment (danialsyed6) [inline src/store/reducers.ts]: create a todo folder in the store folder which should have the actions and reducers of todo.
In that folder create a types folder which should export all the interfaces, types, and enums of todo.

File: src/store/reducers.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    case "REMOVE_TODO":
+      return {
+        todos: state.todos.filter((_, index: number) => index != action.payload),
+      };
+    case "EDIT_TODO":
+      return {
+        todos: state.todos.map((todo: string, index: number) =>
+          index == action.payload.index ? action.payload.value : todo
+        ),
+      };
```
Comment (danialsyed6) [inline src/store/reducers.ts]: use enums for these types

File: src/components/TodoItem/TodoItem.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import React, { useState } from "react";
+import styles from "@/components/TodoList/TodoList.module.css";
+
+type todoItemProps = {
+  keyId: number;
+  value: number;
+  editListHandler: (index: number, value: string) => void;
+  removeListHandler: (index: number) => void;
+};
```
Comment (JavedJutt) [inline src/components/TodoItem/TodoItem.tsx]: create interfaces rather than props. and use as ITodoItemProps

File: src/store/reducers.ts (hunk 5 above)
Comment (danialsyed6) [inline src/store/reducers.ts]: Todo state should also have a loading and error state along with the todos. Set the loading to true when get todos is called and error to true if there is an error in the query.
If you are not using a query then use json server for fake api.
When the loading state is true show a loader and when the error state is true show an error message and clear the error state.

File: src/components/TodoItem/TodoItem.tsx (hunk 1 above)
Comment (JavedJutt) [inline src/components/TodoItem/TodoItem.tsx]: you can just do `setIsCompleted(isChecked)`

Comment (JavedJutt) [reply src/components/TodoItem/TodoItem.tsx]: or you can use better variable name that allies with setIsCompleted

File: src/store/actions.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+});
+
+export const removeTodo = (id: number) => ({
+  type: "REMOVE_TODO",
+  payload: id,
+});
+export const editTodo = (editTaskIndex: number, editTaskValue: string) => ({
+  type: "EDIT_TODO",
+  payload: { index: editTaskIndex, value: editTaskValue },
+});
```
Comment (danialsyed6) [inline src/store/actions.ts]: also create getTodos, clearError, and getTodosFail to set the error state.

File: src/components/TodoList/TodoList.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import React, { useState } from "react";
+import styles from "@/components/TodoList/TodoList.module.css";
+import TodoItem from "@/components/TodoItem/TodoItem";
+
+type TodoListProps = {
+  todos: string[];
+  addTask: (text: string) => void;
+  removeTask: (index: number) => void;
+  editTask: (index: number, value: string) => void;
+};
```
Comment (JavedJutt) [inline src/components/TodoList/TodoList.tsx]: same here for interfaces

File: src/components/TodoList/TodoList.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        type="text"
+        placeholder="Input task name and press enter to add"
+        onChange={(event) => setTaskInput(event.target.value)}
+        onKeyDown={(event) => addListHandler(event.key)}
+      />
+      <hr />
+      <ul className={styles.listContainer}>
+        {todos.map((val: string, i: number) => (
+          <TodoItem
+            key={i}
```
Comment (JavedJutt) [inline src/components/TodoList/TodoList.tsx]: use index rather than i or meaningfull name

File: src/containers/TodoListContainer/index.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable react-refresh/only-export-components */
+import { connect } from "react-redux";
+import TodoList from "@/components/TodoList/TodoList";
+import { addTodo, removeTodo, editTodo } from "@/store/actions";
+
+const mapStateToProps = (state: object) => ({
+  todos: state.todos,
+});
```
Comment (JavedJutt) [inline src/containers/TodoListContainer/index.tsx]: what is object?

File: src/store/store.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { createStore } from "redux";
+import rootReducer from "@/store/reducers";
```
Comment (danialsyed6) [inline src/store/store.ts]: this is not a root reducer this is just the todo reducer. create a root reducer file which should create a root reducer that can join todo reducer with other reducers.

File: src/store/reducers.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+interface TodoState {
+  todos: string[];
+}
+
+const initialState: TodoState = {
```
Comment (JavedJutt) [inline src/store/reducers.ts]: use interfaces start with I and then Name for inteface

File: src/store/reducers.ts (hunk 4 above)
Comment (JavedJutt) [inline src/store/reducers.ts]: ditto

File: src/store/reducers.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  todos: [],
+};
+interface RootState {
+  todo: TodoState;
+}
+
+const todoReducer = (state = initialState, action: any) => {
+  switch (action.type) {
```
Comment (JavedJutt) [inline src/store/reducers.ts]: why any? try to typecast it

**Review by danialsyed6 — CHANGES_REQUESTED** [review]
Implement this with json server fake api

**Pull request merged**